<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Porthoda by T-B-F</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Porthoda</h1>
      <h2 class="project-tagline">python wrapper using protein domain to speed up proteinortho</h2>
      <a href="https://github.com/T-B-F/porthoda" class="btn">View on GitHub</a>
      <a href="https://github.com/T-B-F/porthoda/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/T-B-F/porthoda/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h2>
<a id="requirement-the-following-softwares-and-database-must-be-installed" class="anchor" href="#requirement-the-following-softwares-and-database-must-be-installed" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Requirement, the following softwares and database must be installed</h2>

<ul>
<li>python 2.7</li>
<li>pfam_scan.pl</li>
<li>hmm database pfam-27.0</li>
<li>hmmscan </li>
<li>proteinortho (version 4.0, 5.+ not supported)</li>
<li>libboost</li>
</ul>

<h2>
<a id="method" class="anchor" href="#method" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Method</h2>

<p>porthoDom is a wrapper around proteinortho version 4.26.
Latter version are currently not supported but we are working on it.</p>

<p>porthoDom performs the domain annotation of your protein sequences or can use 
pre-existing annotation in pfam format</p>

<p>The domain arrangements present in the input protein sequences set are clustered
together based on our new domain similarity distance.</p>

<p>The clusters are used as sub-space of search for orthologous candidates, ie
proteins sequences with similar domain arrangement are grouped together per 
species and proteinortho is runned on these sub-groups.</p>

<p>The results of the different proteinortho runs are gathered together and a 
standard proteinortho output file is created.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h2>

<p>porthoDom is composed of two parts, a C program computing the pairwise domain
similarity and a python gluing everything together.</p>

<h3>
<a id="for-a-local-installation" class="anchor" href="#for-a-local-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>For a local installation</h3>

<pre><code>python install setup.py --user
</code></pre>

<p>Be sure that the local installation path is in our python environment path
add  $HOME/.local/lib in PYTHONPATH</p>

<pre><code>export PYTHONPATH=$HOME/.local/lib/
</code></pre>

<p>add the previous line to your bashrc or personal bashrc</p>

<pre><code>echo "export PYTHONPATH=$HOME/.local/lib/" &gt;&gt; ~/.bashrc
</code></pre>

<p>move the local compute_similarity program from bin to a directory where it will
be accessible by the system</p>

<h3>
<a id="if-administrator-right-are-provided" class="anchor" href="#if-administrator-right-are-provided" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>If administrator right are provided</h3>

<pre><code>sudo python install setup.py
</code></pre>

<h2>
<a id="how-to-cite" class="anchor" href="#how-to-cite" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>How to cite</h2>

<p>Domain similarity based orthology detection. <strong>Tristan Bitard-Feildel</strong>, Carsten Kemena, Jenny M Greenwood and Erich Bornberg-Bauer. <strong>BMC Bioinformatics</strong> 2015 16:154</p>

<p><a href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0570-8">BMC paper</a></p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/T-B-F/porthoda">Porthoda</a> is maintained by <a href="https://github.com/T-B-F">T-B-F</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
